### PR TITLE
If an integer is entered with an upper-case base prefix (0Xbeef, 0O755, 0B1010), suggest to make it lowercase

### PR DESCRIPTION
**Pull request by ghost**
The current error for this case isn't really great, it just complains about the whole thing past the `0` being an invalid suffix.

### CODE REVIEW

File: compiler/rustc_parse/src/parser/expr.rs [hunk 1]
```diff
@@ -1718,6 +1731,19 @@ impl<'a> Parser<'a> {
                     self.struct_span_err(span, &msg)
                         .help("valid widths are 8, 16, 32, 64 and 128")
                         .emit();
+                } else if let Some(fixed) = fix_base_capitalisation(suf) {
+                    let msg = format!("invalid suffix `{}` for number literal", suf);
```
Comment (wesleywiser) [inline compiler/rustc_parse/src/parser/expr.rs:1735]:
```suggestion
                    let msg = format!("invalid prefix `{}` for number literal", suf);
```

File: compiler/rustc_parse/src/parser/expr.rs [hunk 2]
```diff
@@ -1718,6 +1731,19 @@ impl<'a> Parser<'a> {
...
+
+                    self.struct_span_err(span, &msg)
+                        .span_label(span, format!("invalid suffix `{}`", suf))
```
Comment (wesleywiser) [inline compiler/rustc_parse/src/parser/expr.rs:1738]:
```suggestion
                        .span_label(span, format!("invalid prefix `{}`", suf))
```

File: compiler/rustc_parse/src/parser/expr.rs (hunk 1 above)
Comment (wesleywiser) [reply compiler/rustc_parse/src/parser/expr.rs:1735]: We might want this to be ``invalid prefix `0{}` for number`` I guess depending on how the error message looks.

**Comment by wesleywiser** [comment]
This is really nice! I think the main thing we need are some [UI tests](https://rustc-dev-guide.rust-lang.org/tests/adding.html) to cover these cases. Let me know if you need any help with that! 🙂

**Comment by 5225225** [comment]
Thanks!

I mainly made this now so I can't then forget about it. I've left PRs in progress for Far Too Long a few times.

I'll make the wording changes you suggested, then add tests for it, and then it should be good to go.

I was debating if I should only make this suggestion if the resulting literal would parse, but will probably decide against that for two reasons

1. That's hard. Or, at least, I couldn't find any way to Just Parse a &str into a Lit in a way that doesn't involve reimplementing the numeric parser but worse.
2. Going from `0B101I` to `0b101I` makes the error message more useful, even if it's still an error.

**Comment by 5225225** [comment]
UI tests here don't work and it's too late in the night for me to work out why the note is considered expected, lol.

Will look at the error annotation syntax some more and hopefully get this done over the weekend.

**Comment by 5225225** [comment]
Okay, bikeshed the error messages with some people and came up with this.

Marking this as ready to review because I'm happy with the diagnostic as-is.

One difference between current and previous is that the previous diagnostic included the actual number in the `error:` text. I figured that it's not needed, since we're pointing to it in a span.

**Review by wesleywiser — APPROVED** [review]
This looks great to me! I'd just like to see the test expanded for the cases I mentioned before we merge.

File: src/test/ui/numeric/uppercase-base-prefix.rs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    //~| HELP try making the prefix lowercase
+    //~| SUGGESTION 0b10101010
+
+    let d = 0XABCDEF_u64;
+    //~^ ERROR invalid base prefix for number literal
+    //~| NOTE base prefixes (`0xff`, `0b1010`, `0o755`) are lowercase
+    //~| HELP try making the prefix lowercase
+    //~| SUGGESTION 0xABCDEF_u64
+
+    let e = 0O755_u32;
```
[the remark this answers is not in the document]
Comment (5225225) [reply src/test/ui/numeric/uppercase-base-prefix.rs:30]: Okay, also added tests for underscores in the middle of the literal (both with/without type suffixes).

**Comment by wesleywiser** [comment]
Thanks!

@bors r+ rollup

**Pull request merged**